### PR TITLE
[Major] Feature/axios and angular abstraction

### PR DESCRIPTION
**Pull request by SimonSchick**
This decouples the `APIAdapter` from axios and adds a new http client (angular).

Breaking changes:
- ~needs angularv4 (can downgrade, but I want to wait for v4 to become main for us)~ this happened :mappa:
- axios no longer a fixed depedency
- Carina no longer a fixed dependency
- http client now needs to be passed into api adapter constructor
- All raw request APIs have been abstracted, they should be compatible for the most part but I expect some changes required on fe2.
  - All header values are now arrays (this is more consistent)
  - status => statusCode
  - data => body (debatable)
  - params => query (is clearer that way)
- Removal of status text as it's pretty useless (debatable)

### CODE REVIEW

**Comment by SimonSchick** [comment]
Ignore travis for now, they broke their site lol.

**Comment by SimonSchick** [comment]
@JamyDev

File: src/adapters/http/HTTPInterface.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    url: string;
+    method?: 'post' | 'put' | 'get' | 'head' | 'delete' | 'patch';
+    body?: string | {} | FormData;
+    query?: { [param: string]: StringIsh };
+    headers?: { [header: string]: StringIsh[] };
+    json?: boolean;
+    withCredentials?: boolean;
+}
+
+export interface IRequestResponse {
```
Comment (JamyDev) [inline src/adapters/http/HTTPInterface.ts]: Rename this to `IResponse` to avoid confusion.

**Comment by SimonSchick** [comment]
@connor4312 poke

File: src/adapters/APIAdapter.ts [hunk 2]
```diff
@@ -242,23 +242,23 @@ export class APIAdapter implements IPersistentAdapter {
...
+                query: this.queryBuilder.build(resolveOptions),
             };
             if (options.rawOptions) {
                 merge(request, options.rawOptions);
             }
             return this.request(request, model);
         })
         .then(response => {
-            if (!(response.data instanceof Array)) {
+            if (!(response.body instanceof Array)) {
```
Comment (UnwrittenFun) [inline src/adapters/APIAdapter.ts]: Array.isArray

Comment (SimonSchick) [inline src/adapters/APIAdapter.ts]: Any reason to prefer that?

Comment (SimonSchick) [inline src/adapters/APIAdapter.ts]: Any reason to prefer that?

Comment (UnwrittenFun) [inline src/adapters/APIAdapter.ts]: It's more robust (works across iframes, though admittedly a niche situtation). Plus it reads better 😝

Comment (SimonSchick) [inline src/adapters/APIAdapter.ts]: iframes? Can you elaborate?

Comment (SimonSchick) [inline src/adapters/APIAdapter.ts]: I think I will leave it as is, we use it plenty elsewhere 😛

File: src/adapters/APIAdapter.ts [hunk 3]
```diff
@@ -293,20 +293,20 @@ export class APIAdapter implements IPersistentAdapter {
...
             merge(req, options.rawOptions);
         }
         return this.request(req, model.ctor)
         .then(response => {
-            if (response.status === 204) { // some endpoints do not return the created resource
+            if (response.statusCode === 204) { // some endpoints do not return the created resource
                 return model;
             }
-            model.extend(response.data, {
+            model.extend(<{}>response.body, {
```
Comment (ProbablePrime) [inline src/adapters/APIAdapter.ts]: Could you use `IJSON` here instead of `{}`

File: src/adapters/errors/AccessError.ts [hunk 4]
```diff
@@ -10,13 +10,19 @@ export class AccessError extends AdapterResponseError {
...
+        response: IResponse,
         adapter: IPersistentAdapter,
         model?: IModelCtor<IModel>,
     ) {
         super(response, adapter, model);
         AccessError.setProto(this);
-        this.scope = response.data.scope;
-        this.resource = response.data.resource;
+
+        const {
```
Comment (ProbablePrime) [inline src/adapters/errors/AccessError.ts]: This reads poorly but makes sense, anything we can do here?

Comment (SimonSchick) [inline src/adapters/errors/AccessError.ts]: It reads fine for me, not really sure how we can change that :/

File: src/adapters/APIAdapter.ts (hunk 3 above)
Comment (SimonSchick) [inline src/adapters/APIAdapter.ts]: Can probably use `object` now.

**Pull request merged**